### PR TITLE
Advent 2016: Cloudforest post

### PR DESCRIPTION
**Pull request by vdemario**
Sorry for the delay!

I'm not quite comfortable with the shameless self promotion of my GopherCon Brasil talk in the end of the post and the link to Mendelics (the company I work for) in the genetics section but both of those felt like they made sense in their respective sections. Let me know you think I should remove them.

Any feedback and/or edits is much appreciated. I may answer faster if you ping me on Slack, I'm on the #writing channel, same username (vdemario).

cc @ryanbressler

### CODE REVIEW

File: content/advent-2016/cloudforest.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+series = ["Advent 2016"]
++++
+
+[CloudForest](https://github.com/ryanbressler/CloudForest) is a machine learning project dedicated to the construction of [Random Forests](https://www.stat.berkeley.edu/~breiman/RandomForests/cc_home.htm) built entirely in Go. It was created by [Ryan Bressler](https://github.com/ryanbressler).
+
+Random Forests are a machine learning algorithm based around the construction of many single classification trees, each splitting both the training set and the features available to train the model randomly. Each single tree is different from the others due to this random split and the ensemble of all the trees together is able to classify the data better than any single tree could do by itself. We won't go deep on how Random Forests work internally, you can learn more [here](https://www.stat.berkeley.edu/~breiman/RandomForests/cc_home.htm#intro).
+
+CloudForest is a swiss army knife of options for how a model can be built. It supports not only the original ideas from Breiman and Cutler's algorithm but also boosting, class weighted classification, feature selection with artificial constrasts, support for missing values and many more features.
```
Comment (bradleyfalzon) [inline content/advent-2016/cloudforest.md]: Should `artificial constrasts` be `artificial constraints`?

File: content/advent-2016/cloudforest.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+# Extra options for growforest
+
+To support all of the different ways CloudForest can build classification models, there are dozens of flags. The list of flags for all the options can feel daunting when you're not familiar with some of the terms, much like this:
+
+![CloudForest options](/postimages/advent-2016/cloudforest/controlpanel.png)
+
+However, some of the flags are of particular importance to any kind of model:
+
+- `nTrees`: the number of trees. By default, if you don't specify anything for this option, the forest will be built with `-nTrees=100`. A hundred trees if often a good enough number, but there is no one number of trees that will be appropriate for every situation.
```
Comment (bradleyfalzon) [inline content/advent-2016/cloudforest.md]: Should `A hundred trees if often` be `A hundred trees is often`?

File: content/advent-2016/cloudforest.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+- `mTry`: the number of candidate features to be selected with each tree. Not every tree in a forest is build the same. Each classification tree is built with a subset of the features. `-mTry` determines how many features are available for selection during the construction of each tree. Lower values yield trees with less correlation and higher values allow more opportunities for important features to be selected. If not specified, `-mTry` is the square root of the number of features. This value is common across many implementations of Random Forests, and it's one of the parameters that has the highest impact on the accuracy of the final model. You may want to experiment with bigger values for `-mTry`. We've had success with much higher ratios, such as .5 (50% of the features available for each tree), but your mileage may vary.
```
Comment (bradleyfalzon) [inline content/advent-2016/cloudforest.md]: Should `is build` be `is built`, but then it seems a little repetitive with the sentence immediately after.

File: content/advent-2016/cloudforest.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+- `maxDepth`: depending on your data, some trees may have a hard time converging. `-maxDepth` determines how tall a tree can grow before the algorithm cuts it. Without `-maxDepth`, `-growforest` can run fast at first, completing the construction of several trees and then slow down in the end when some of the feature sets have trouble finding a good way to split the cases.
+
+# Importing and using CloudForest directly from other Go programs
+
+CloudForest was built as a Go package. We've seen only command line utilities so far, but these command line utilities are just `main` packages that import the base package and use its functions. Any other Go program can do the same.
+
+For example, loading an AFM file into memory can be done like this:
+
+
+```
```
Comment (bradleyfalzon) [inline content/advent-2016/cloudforest.md]: You can add the keyword `go` here (and obviously other blocks further down) to get syntax highlighting.

File: content/advent-2016/cloudforest.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+			name := strings.ToLower(structField.Name)
+			field := fmt.Sprintf("%s:%s", prefix, name)
+			fields = append(fields, field)
+		}
+	}
+	return fields
+}
+```
+
+Forest construction is still done in the command line with `growforest`. Application of the model to new, unknown mutations, is done entirely in memory, as the system receives new files for processing. At the time this was originally built, without much experience in Go, we did not realize that all of this work was actually the construction of an encoding library. It does not conform to the encoding interfaces in the standard library due to that oversight and has never seen the light of open source, sadly.
```
Comment (bradleyfalzon) [inline content/advent-2016/cloudforest.md]: nit: should `in the command line` be `on the command line`?

File: content/advent-2016/cloudforest.md [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+In production, we've had multiple models built with `growforest` and applied this way, always achieving more than 90% accuracy accross all of our tests. The system can't predict by itself every single mutation that causes a disease in real patients but serves as a powerful tool for the geneticists to quickly find what really matters.
```
Comment (bradleyfalzon) [inline content/advent-2016/cloudforest.md]: `accross` should be `across`.

File: content/advent-2016/cloudforest.md (hunk 1 above)
Comment (ryanbressler) [inline content/advent-2016/cloudforest.md]: Contrasts is correct...maybe link to the paper http://www.jmlr.org/papers/volume10/tuv09a/tuv09a.pdf

Comment (bradleyfalzon) [inline content/advent-2016/cloudforest.md]: oh til, a link would be great, thank you

**Comment by ryanbressler** [comment]
Looks great!

I'd definetly leave in the links to your talk and company.

If you want to you could include a refence to http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0144820 it talks about some of the reasons cloudforest works so well on genomic data.

File: content/advent-2016/cloudforest.md (hunk 1 above)
Comment (vdemario) [inline content/advent-2016/cloudforest.md]: Yes, that's right. I've added a link.

File: content/advent-2016/cloudforest.md (hunk 3 above)
Comment (vdemario) [inline content/advent-2016/cloudforest.md]: Yes, it should. I tweaked the following sentence too, it should be better now.

File: content/advent-2016/cloudforest.md (hunk 4 above)
Comment (vdemario) [inline content/advent-2016/cloudforest.md]: Great! I've done it for all the blocks. It helped me catch a mistake in the first block, too.

File: content/advent-2016/cloudforest.md (hunk 5 above)
Comment (vdemario) [inline content/advent-2016/cloudforest.md]: Yeah, `on the command line` sounds betters. Fixed it.

File: content/advent-2016/cloudforest.md (hunk 6 above)
Comment (vdemario) [inline content/advent-2016/cloudforest.md]: Fixed it, thanks.

**Comment by vdemario** [comment]
Thanks for the reviews @bradleyfalzon and @ryanbressler. I've pushed several fixes based on your feedback.

@ryanbressler I've linked to the CloudForest paper discreetly, from the _impressive results_ sentence. Let me know if you think it needs to be more visible, I could a sentence about it.

**Pull request merged**